### PR TITLE
Respect promises on pipelined responses

### LINKED ISSUE #23322: Tests: Netty4PipeliningEnabledIT#testThatNettyHttpServerSupportsPipelining fails reproducibly
On current master (771fd1f4eaa700ddecf634dee04be80f0a40471f): https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+g1gc/1844/console

'''
gradle :modules:transport-netty4:integTestRunner -Dtests.seed=BF593F0EC855DB4E -Dtests.class=org.elasticsearch.http.netty4.Netty4PipeliningEnabledIT -Dtests.method="testThatNettyHttpServerSupportsPipelining" -Dtests.security.manager=true -Dtests.jvm.argline="-XX:-UseConcMarkSweepGC -XX:+UseG1GC" -Dtests.locale=ko -Dtests.timezone=Asia/Beirut
'''

### PR DESCRIPTION
**Pull request by jasontedor**
When pipelined responses are sent to the pipeline handler for writing, they are not necessarily written immediately. They must be held in a priority queue until all responses preceding the given response are written. This means that when write is invoked on the handler, the promise that is attached to the write invocation will not necessarily be the promise associated with the responses that are written while the queue is drained. To address this, the promise associated with a pipelined response must be held with the response and then used when the channel context is actually written to. This was introduced when ensuring that the releasing promise is always chained through on write calls lest the releasing promise never be invoked. This leads to many failing test cases, so no new test cases are needed here.

Relates #23310, closes #23322

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java [hunk 1]
```diff
@@ -33,7 +33,8 @@
  */
 public class HttpPipeliningHandler extends ChannelDuplexHandler {
 
-    private static final int INITIAL_EVENTS_HELD = 8;
```
Comment (bleskes) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java]: Is this related to this change? if not , should we move it to separate PR?

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java [hunk 2]
```diff
@@ -80,14 +78,19 @@ public void write(final ChannelHandlerContext ctx, final Object msg, final Chann
...
                             break;
                         }
                         holdingQueue.remove();
-                        ctx.write(response.response(), promise);
+                        /*
+                         * We must use the promise attached to the response; this is necessary since are going to hold a response until all
+                         * responses that precede it in the pipeline are written first. Note that the promise from the method invocation is
+                         * not ignored, it will already be attached to an existing response and consumed when that response is drained.
+                         */
+                        ctx.write(response.response(), response.promise());
```
Comment (bleskes) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java]: Should we assert at the beginning of this if clause that the promise in the HttpPipelinedResponse is that one we got as a parameter?

**Review by bleskes — APPROVED** [review]
LGTM, though I don't feel comfortable I can oversee unexpected side effects (thanks for slowly making me learn netty). Maybe @spinscale can look at it as well?

**Comment by spinscale** [comment]
looks good to me

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java (hunk 1 above)
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java]: Sure, I'll move it to a separate change.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java (hunk 2 above)
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java]: I pushed 072c81d1cc9b420ba022301592f7e76fa013b313.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java (hunk 1 above)
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java]: I pushed 70d57b75689e06a286715818968b069e01ed6648.

**Comment by jasontedor** [comment]
Thanks for reviewing @bleskes and @spinscale.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java (hunk 1 above)
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java]: I opened #23335.

**Pull request merged**